packages/ui-utils/src/is-object-interpolation/index.js

```diff
@@ -0,0 +1,49 @@
+/* global TemplateStringsArray */
+/**
+ * External dependencies
+ */
+import isPlainObject from 'lodash/isPlainObject';
+
+/* eslint-disable jsdoc/valid-types */
+/**
+ *
+ * There's no way to detect whether an `any` type is an `ObjectInterpolation` because `ObjectInterpolation`s
+ * are merely objects where the keys are either (1) [CSS property names](https://github.com/emotion-js/emotion/blob/a72e6dc0f326b7d3d6067698d433018ee8c4cbf1/packages/serialize/types/index.d.ts#L10),
+ * (2) [CSS Pseudo property names](https://github.com/emotion-js/emotion/blob/a72e6dc0f326b7d3d6067698d433018ee8c4cbf1/packages/serialize/types/index.d.ts#L18),
+ * or (3) [anything else where the value for the key is another interpolation](https://github.com/emotion-js/emotion/blob/a72e6dc0f326b7d3d6067698d433018ee8c4cbf1/packages/serialize/types/index.d.ts#L19).
+ * That means ObjectInterpolations can look all of the following ways:
+ *
+ * ```js
+ * const oi = {
+ *   color: 'black',
+ * };
+ *
+ * const oi = {
+ *   ':disabled': css`color: black;`,
+ * };
+ *
+ * const oi = {
+ *   '@media (prefers-color-scheme: light)': css({
+ *     backgroundColor: 'black',
+ *     color: 'white',
+ *   }),
+ * };
+ * ```
+ *
+ * Therefore, this function only accepts `TemplateStringsArray` and `Interpolation` for it's arguments rather than `any`, which
+ * cannot be refined to an `ObjectInterpolation`. However, given any generic `Interpolation`, we _can_ tell whether it is an
+ * `ObjectInterpolation` specifically by simply checking whether the value is "plain object" as defined by `lodash`'s `isPlainObject`.
+ *
+ * Note: The links above reference a specific commit for **Emotion 10**. Emotion 11 is out but we're stuck on Emotion 10 until we can
+ * re-write `create-styles` and upgrade all of `@wordpress/components` to use the new version of Emotion.
+ *
+ * At that point we'll have to revisit this function anyway as it seems that the `ObjectInterpolation` type no longer
+ * exists and that the concept has evolved.
+ *
+ * @param {TemplateStringsArray | import('create-emotion').Interpolation} value The value to check.
+ * @return {value is import('create-emotion').ObjectInterpolation} Whether the value is an ObjectInterpolation.
+ */
+const isObjectInterpolation = ( value ) => isPlainObject( value );
```

Comment:
why not just use isPlainObject directly?

Reply:
The usefulness of this function comes primarily from the type defintions. It's impossible to refine the type any other way other than through type definitions which TypeScript will trust implicitly. This allows the newly added packages to be fully and completely typed.

For example, given the following (real) code example:

```js
/**
 * @param {TemplateStringsArray | import('create-emotion').Interpolation<undefined>} template
 * @param {import('create-emotion').Interpolation<undefined>[]} args The styles to compile.
 * @returns {ReturnType<compile>} The compiled styles.
 */
export function css(template, ...args) {
	if (isObjectInterpolation(template)) {
		return compile(responsive(template));
	}

	if (Array.isArray(template)) { // => TemplateStringsArray
		for (let i = 0, len = template.length; i < len; i++) {
			const n = template[i];
			if (isObjectInterpolation(n)) {
				template[i] = responsive(n);
			}
		}
		return compile(template, ...args);
	}

	return compile(template, ...args);
}
```

If you replace `isObjectInterpolation` with `isPlainObject` and `responsive`'s first argument will not be refined enough to be accepted as `responsive` only accepts an `ObjectInterpolation` for it's first argument.

Reply:
I don't think it's a valid reason for us at least to add a new package to be honest. Can't this be solved by an explicit type cast instead?

Reply:
Well, this package isn't complete as it stands, so it's justification can't live on just these two utilities. The issue linked in the PR description has a short list of utilities that would live in this package (and those are just the ones used by the `Text` and `Truncate` components, there are many more if you check out the G2 repo).

FWIW I hadn't tried to type cast it but it does work, so we can remove this utility if you'd prefer to just use type casts (that's all this function does anyhow).

Regardless, I'll open a separate PR with a different utility adding the same `ui-utils` package for `getOptimalTextShade`, `getDisplayName`, and `getComputedColor` as the issue linked in the PR describes. This was just meant as an incremental way of doing it without moving more code at one time than is possible for a person to effectively review.

Reply:
I think adding packages shouldn't be seen as something we should do easily especially in the context of WordPress, so I'd personally feel better if we start by doing inline utilities in the "components" package and discuss whether it deserves a separate package or not as the usage grows.

Reply:
What about utilities used by mulitple packages like `getDisplayName`? Where should that live? We can't introduce them in `components` as that would create a circular dependency. We can duplicate the code in the packages if you'd prefer that. Or we can export them from an unrelated package, but then that package will be responsible for something that it shouldn't care about.

Reply:
What does `getDisplayName` do?

Reply:
```js
/**
 * Gets the displayName of a React Component or element.
 *
 * @param {string | import('react').ComponentType} tagName
 *
 * @return {string} The display name of the Component / tagName.
 */
export function getDisplayName(tagName) {
	const displayName =
		typeof tagName === 'string'
			? tagName
			: tagName.displayName || tagName.name || 'Component';

	return displayName;
}
```

Reply:
So looking a g2. It seems to be only used once. That said it seems to be used to compute components display name? In fact we have something close to it in `createHigherOrderComponent` in `@wordpress/compose` so  it does seem like a natural place for it if its usage grows.

Reply:
Ah, you're right, I thought `context` used it as well but I was wrong. We'll inline that one too I guess.